src/main/java/org/logstash/beats/BatchEncoder.java

```diff
@@ -0,0 +1,69 @@
+package org.logstash.beats;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.MessageToByteEncoder;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.io.IOException;
+import java.util.Map;
+
+
+public class BatchEncoder extends MessageToByteEncoder<Batch> {
+    private final static Logger logger = LogManager.getLogger(BatchEncoder.class.getName());
+
+
+    @Override
+    protected void encode(ChannelHandlerContext ctx, Batch batch, ByteBuf out) throws Exception {
+        out.writeByte(batch.getProtocol());
+        out.writeByte('W');
+        out.writeInt(batch.size());
+        out.writeBytes(getPayload(ctx, batch));
+    }
+
+    protected ByteBuf getPayload(ChannelHandlerContext ctx, Batch batch) throws IOException {
+        ByteBuf payload = ctx.alloc().buffer();
```

Comment:
and here begin the nitty gritty details. By default (please use a debugger or print out `payload.isDirect()` to verify) this allocates a direct pooled byte buffer in netty 4.1 - which means it is outside the JVM and not in GC reach (and by default it is unbounded IIRC). So this requires some monitoring during testing. You can use `-XX:MaxDirectMemorySize` as jvm start up arg to limit this. Also visualvm allows you to install a `buffer pools` plugin which monitors direct memory

also, if you know the size of the batch, you can allocate the buffer into the right size